scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java

```diff
@@ -0,0 +1,182 @@
+package gov.nasa.jpl.aerie.scheduler;
+
+import gov.nasa.jpl.aerie.merlin.driver.ActivityInstanceId;
+import gov.nasa.jpl.aerie.merlin.driver.MissionModel;
+import gov.nasa.jpl.aerie.merlin.driver.SerializedActivity;
+import gov.nasa.jpl.aerie.merlin.driver.SimulationResults;
+import gov.nasa.jpl.aerie.merlin.driver.engine.SimulationEngine;
+import gov.nasa.jpl.aerie.merlin.driver.engine.TaskId;
+import gov.nasa.jpl.aerie.merlin.driver.timeline.LiveCells;
+import gov.nasa.jpl.aerie.merlin.driver.timeline.TemporalEventSource;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class IncrementalSimulationDriver {
+
+  private Duration curTime;
+  private SimulationEngine engine;
+  private LiveCells cells;
+  private TemporalEventSource timeline;
```

Comment:
I'd suggest making these `final` and refactoring this class to use some sort of builder pattern.

These 4 members are initialized to `null` by default and there's an assumption that an assignment occurs before their first real use. That's usually not too complicated to get right in your head but when other people have to maintain that logic it's easy for `NullPointerException`s to arise out of forgetfulness.

Reply:
Basically I try to avoid "lazily assigned" variables (non-`final` and without a default) as much as possible.

Reply:
These four are assigned each time we reset the simulation, which is not controlled by the user. I do not see how to make them final. Could you elaborate ?

Reply:
Ahh I see. I'd at least change this to:
```java
this.engine = new SimulationEngine();
  private Duration curTime = Duration.ZERO;
  private SimulationEngine engine = new SimulationEngine();
  private TemporalEventSource timeline = new TemporalEventSource();
```
to avoid `null` defaults. 

Another idea would be to make a `static` inner-class here called `Builder` to store all of the variables that are non-`final`. This way class members are at least organized by mutability/immutability and referring to any mutable variable would require qualifying with a `builder.*`. More of a way to signal to future maintainers that these variables are mutable rather than creating anything really enforceable, but better than no signal maybe.

Reply:
Hmm... is anything being built here? I like your idea about grouping them in a static inner class, I'm just not sure I'd use the word "builder". Would something like this be similar to what you're describing, @pcrosemurgy?
```
record SimulationDetails(Duration curTime, SimulationEngine engine, TemporalEventSource timeline)
```

Reply:
I am not so convinced
- because `cells` needs a `TemporalSource` and the mission model, we cant' do anything about it
- we double initialize these variables when creating the object (default initializers + `initSimulation()`) 

Some classes are stateless, this class is very stateful and `initSimulation()` makes it clear and provides an initialization that has some meaning. 
 
I will put default initializers as you first suggested if that's fine.

Reply:
That's fair, thanks Adrien!

Reply:
Thank you !